### PR TITLE
add pcm solvent models

### PR DESCRIPTION
**Pull request by wxj6000**
- added PCM family solvent models including C-PCM, COSMO, IEF-PCM, SS(V)PE
- added their analytical gradients

### CODE REVIEW

File: pyscf/solvent/pcm.py [hunk 1]
```diff
@@ -0,0 +1,627 @@
...
+        logger.info(self, 'eta = %s'          , self.eta)
+        logger.info(self, 'eps = %s'          , self.eps)
+        logger.info(self, 'frozen = %s'       , self.frozen)
+        logger.info(self, 'equilibrium_solvation = %s', self.equilibrium_solvation)
+        logger.debug2(self, 'radii_table %s', self.radii_table)
+        if self.atom_radii:
+            logger.info(self, 'User specified atomic radii %s', str(self.atom_radii))
+        self.grids.dump_flags(verbose)
+        return self
+
```
Comment (sunqm) [inline pyscf/solvent/pcm.py:502]: When molecule object is changed, the surface should be updated accordingly, right? Here a `reset` method should be add to clear up the contents of `surface`, and also recursively call the `super().reset()` .

File: pyscf/solvent/pcm.py [hunk 2]
```diff
@@ -0,0 +1,627 @@
...
+        solvent_obj = PCM(mf.mol)
+    return _attach_solvent._for_scf(mf, solvent_obj, dm)
+
+
+# Inject PCM to other methods
+from pyscf import scf
+from pyscf import mcscf
+from pyscf import mp, ci, cc
+from pyscf import tdscf
+scf.hf.SCF.PCM    = scf.hf.SCF.PCM    = pcm_for_scf
```
Comment (sunqm) [inline pyscf/solvent/pcm.py:48]: To make the attribute PCM registered in SCF class, the pcm module can be imported in `solvent/__init__.py`

File: pyscf/solvent/pcm.py [hunk 3]
```diff
@@ -0,0 +1,627 @@
...
+    if solvent_obj is None:
+        solvent_obj = PCM(mf.mol)
+    return _attach_solvent._for_scf(mf, solvent_obj, dm)
+
+
+# Inject PCM to other methods
+from pyscf import scf
+from pyscf import mcscf
+from pyscf import mp, ci, cc
+from pyscf import tdscf
```
Comment (sunqm) [inline pyscf/solvent/pcm.py:47]: These modules have not been used in this module. They should not be imported here.

File: pyscf/solvent/pcm.py [hunk 4]
```diff
@@ -0,0 +1,627 @@
...
+    '''
+    generate F and A matrix in  J. Chem. Phys. 133, 244111 (2010)
+    '''
+    R_vdw = surface['R_vdw']
+    switch_fun = surface['switch_fun']
+    weights = surface['weights']
+    A = weights*R_vdw**2*switch_fun
+    return switch_fun, A
+
+def get_dF_dA(surface):
```
Comment (sunqm) [inline pyscf/solvent/pcm.py:196]: Following gradients code are duplicated from the code in grad/pcm.py . Can they be removed?

File: pyscf/solvent/ddcosmo.py [hunk 5]
```diff
@@ -806,6 +806,7 @@ def _get_vind(self, dm):
             f_epsilon = 1
         epcm = .5 * f_epsilon * numpy.einsum('jx,jx', psi, Xvec)
         vpcm = .5 * f_epsilon * vmat
+        print(dm[:3,:3])
```
Comment (sunqm) [inline pyscf/solvent/ddcosmo.py]: This print is not needed

File: pyscf/solvent/pcm.py [hunk 6]
```diff
@@ -499,6 +278,10 @@ def dump_flags(self, verbose=None):
             logger.info(self, 'User specified atomic radii %s', str(self.atom_radii))
         self.grids.dump_flags(verbose)
         return self
+    
+    def reset(self, mol=None):
+        super().reset(mol)
+        self.surface = None
```
Comment (sunqm) [inline pyscf/solvent/pcm.py:285]: `return self`

File: pyscf/solvent/test/test_ddcosmo_grad.py [hunk 7]
```diff
@@ -876,8 +876,9 @@ def test_ccsd_grad(self):
         mf = scf.RHF(mol2).run()
         mycc2 = solvent.ddCOSMO(cc.CCSD(mf)).run()
         e2 = mycc2.e_tot
+        print((e2-e1)/dx, de[0,2])
```
Comment (sunqm) [inline pyscf/solvent/test/test_ddcosmo_grad.py]: remove the print statement

**Review by sunqm — APPROVED** [review]

**Comment by wxj6000** [comment]
@sunqm can you confirm this is a bug? We are supposed to return self, right?
https://github.com/pyscf/pyscf/blob/master/pyscf/soscf/newton_ah.py#L708

**Comment by sunqm** [comment]
> @sunqm can you confirm this is a bug? We are supposed to return self, right? https://github.com/pyscf/pyscf/blob/master/pyscf/soscf/newton_ah.py#L708

Yes. Your change fixes the bug.

**Pull request merged**